### ISSUE TITLE
Canvas fingerprint changes in Blink on macOS

### ISSUE THREAD

**Issue by Finesse** [fingerprintjs/fingerprintjs#574]
I use Chrome 86 on macOS 10.15.6, my laptop has only 1 GPU. The canvas fingerprint produces one picture on some websites and another on other websites:

| Picture 1 | Picture 2 |
| -------- | ------- |
| ![local](https://user-images.githubusercontent.com/9006227/96951456-dec48780-152f-11eb-8045-ef702db58a49.png) | ![other](https://user-images.githubusercontent.com/9006227/96951466-e1bf7800-152f-11eb-9ba2-f4e02b62380b.png) |

The only difference is the right margin of the emoji. Both pictures are stable: one of them is produced on a website, and their distribution is almost equal.

Use the following browser console script to run get a canvas fingerprint in your browser:

```js
function initFingerprintJS() {
  FingerprintJS.load()
    .then(fp => fp.get())
    .then(result => {
      console.log(FingerprintJS.murmurX64Hash128(result.components.canvas.value.data));
    });
}
var script = document.createElement('script'); script.onload = initFingerprintJS; script.src = 'https://cdn.jsdelivr.net/npm/@fingerprintjs/fingerprintjs@3.0.0/dist/fp.min.js'; document.head.appendChild(script);
```

-----------------

The results are the same in regular and private mode, the zoom level is 100% in all the cases that I've checked.

-----------------

- The difference doesn't happen in Chrome 86 on Windows 10 on my local virtual machine and on Windows 7 and 10 on BrowserStack
- The difference happens in Chrome 86–88 on macOS Catalina and Mojave on BrowserStack
- The difference doesn't happen in Firefox 82 and Safari 14 on my local macOS 10.15.6
- The difference doesn't happen in Chrome 83 on Android 8 (Galaxy S9) and 11 (Pixel 4) on BrowserStack
- The difference happens in other Chromium browsers (Edge, Yandex) on my local macOS 10.15.6

So the issue affects only visitors that use a Chromium-based browser on macOS.

---------------

The problem doesn't affect the version 2, probably because it has no text after the emoji. Run the following code in a browser console to get an open v2 canvas fingerprint:

```js
function initFingerprintJS() {
  Fingerprint2.get((components) => {
    const canvasComponent = components.find(component => component.key === 'canvas');
    console.log(Fingerprint2.x64hash128(canvasComponent.value[1].slice(10)));
  });
}
var script = document.createElement('script'); script.onload = initFingerprintJS; script.src = 'https://cdn.jsdelivr.net/npm/@fingerprintjs/fingerprintjs@2/dist/fingerprint2.min.js'; document.head.appendChild(script);

```

--------------

A fingerprint variant persists until you close the browser tab (even if no fingerprint was taken on the first page of the tab).

-------------

I've found the key difference that causes the image change: at least 1 character on the page must be rendered with the `BlinkMacSystemFont` font, the font can be set through the `font-family` CSS property (the closer the font is to the start of the list, the higher the chance to affect a character). The canvas is affected despite the `BlinkMacSystemFont` font not being used in the canvas image. The fact that the fingerprint variant persists until the tab is closed makes me think that this is a subtle bug of Chromium.

---------------

The right margin of the emoji is reduced regardless of the font that I set to `context.font`. The margin is reduced if the canvas text font size is less than `18pt`, it doesn't depend on these browser settings: font size, page zoom, customize font / font size.

--------------

Possible solutions:
- ~~Ask library users to use the same font-family on every page of their websites or not use `BlinkMacSystemFont` at all (that isn't always possible).~~ It won't help because the picture variant depends on the first page of the browser tab, not on the current page.
- Change the canvas image such way that it isn't affected by this Chrome feature (put the emoji to the end or increate the text size). It will change the fingerprint of all visitors.
- ~~Make the canvas component draw only one of the pictures.~~ I haven't found a way to do so. The picture variant is persistent so the library can affect it only if the agent runs on the first page of the browser tab.
- Address the canvas emoji rendering issue to Google. There is no guarantee that they will fix it and that they will fix it soon.

-------------

We've decided to address the rendering issue to the Chromium team and wait because: 

- The issue affects a small number of users
- A website must meet some special requirements to face this issue

**Comment by Finesse** [comment]
I've found out that the Chromium bug isn't presented when the canvas text size is set via `px` instead of `pt`. @Valve, is there a reason to use `pt` instead of `px`?

**Comment by Finesse** [comment]
This difference was introduced on Chrome 83.

**Comment by spyjo** [comment]
I can confirm this bug is still present in Chrome 87.0.4280.67 on Mac OS 10.15.7.
I just had a case with a website made with Bootstrap, which include BlinkMacSystemFont by default.

**Comment by Finesse** [comment]
TL;DR. The steps to reproduce are:

1. Take the HTML code from https://github.com/fingerprintjs/fingerprintjs/issues/574#issuecomment-714893037 and save it to a file. Upload the file to a web server (optionally).
2. Open Chrome version 83 or newer on macOS, open a **new** tab and open the HTML file in it.
3. Take the HTML code from https://github.com/fingerprintjs/fingerprintjs/issues/574#issuecomment-714893037, **remove** the `font-family: BlinkMacSystemFont;` line and save it to another file. Upload the file to a web server (optionally).
4. Open a **new** tab and open the other HTML file in it.

The images in the tabs will be different.

**Comment by Finesse** [comment]
Fixed by d8370403175c33d2f9542b172f9c9ea9fa3c17bd by moving the emojis to the right of the text lines. The fix will be published in the next big release that combines all changes that breaks the fingerprint compatibility.
